firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -103,4 +124,105 @@ String getApplicationId() {
   String getName() {
     return firebaseApp.getName();
   }
+
+  /**
+   * Create firebase installation id for the {@link FirebaseApp} on FIS Servers and client side
+   * cache.
+   *
+   * <pre>
+   *     The workflow is:
+   *         check if cache empty or cache status is REGISTER_ERROR
+   *                                 |
+   *                           create random fid
+   *                                 |
+   *               update cache with cache status UNREGISTERED
+   *                                 |
+   *                           send http request to backend
+   *                                 |                  |
+   *  on success: set cache entry status to REGISTERED  |
+   *                                 |                  |
+   *                                 |         on failure: set cache entry status to REGISTER_ERROR
+   *                                 |                     |
+   *                                return          throw exception
+   *
+   *
+   *                      else if cached FID exists
+   *                             |            |
+   *     if cache status is UNREGISTERED      |
+   *                  |                   return cached FID
+   *     send http request to backend
+   *               |               |
+   *  on success: set cache        \
+   *  entry status to REGISTERED   \
+   *                |              \
+   *                |         on failure: set cache entry status to REGISTER_ERROR
+   *                |                     |
+   *              return          throw exception
+   * </pre>
+   */
+  @WorkerThread
+  private String createFirebaseInstallationId() throws FirebaseInstallationsException {
+
+    FiidCacheEntryValue cacheEntryValue = localCache.readCacheEntryValue();
+
+    if (cacheEntryValue == null
```

Comment:
I recommend to avoid nested if-clauses.
How about outsourcing the different code paths in named methods and refactor this method like this:

private String getFirebaseInstallationId() {
  if (localFidExistsAndRegistered()) {
    return localFid();
  }

  return generateRegisterSaveAndGetNewFid();
}

Reply:
Updated.